src/middlewared/middlewared/utils/reserved_ids.py

```diff
@@ -0,0 +1,36 @@
+from asyncio import Lock as AsyncioLock
+from dataclasses import dataclass
+from threading import Lock
+from time import monotonic
+from typing import Union
+
+
+LOCKED_XID_TTL = 3600
+
+
+@dataclass(slots=True)
+class ReservedXid:
+    in_flight: dict[int, int]
+    lock: Union[Lock, AsyncioLock]
+
+    def available(self, xid: int) -> bool:
+        if (expiry := self.in_flight.get(xid)) is None:
+            return True
+
+        return monotonic() > expiry
```

Comment:
I know its not the normal case, but if an entry **_has_** expired, should we auto-remove it here?